[macos] add flavor options to commands in the `flutter_tool`

a-wallen opened the pull request:
Adds `--flavor` option to `flutter run` and `flutter build`. Running against preexisting devicelab flavor tests for feature parity between `macOS`, `iOS`, and `Android`. 

It'll be easier to review after https://github.com/flutter/flutter/pull/117352, commits from https://github.com/flutter/flutter/pull/118421/commits/c22f1cc43541c06948441cd4fc51819060c365f1 to https://github.com/flutter/flutter/pull/118421/commits/727ee2c1135cc59ea2afd36cb659b9695d329b19 are from that PR.

Fixes https://github.com/flutter/flutter/issues/64088 

## Flavors `macOS` Checklist
- [x] `flutter build`
- [ ] ~`flutter install`~ not needed since `install` doesn't do any work on `DesktopDevice`s
- [x] `flutter run`  
- [x] `flutter drive`

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

Commit:
Add new macos target configured for flavors

Commit:
Pipe flavor as scheme into xcodebuild

Commit:
Ignore incoming flavor string

Commit:
Rename Free App copy-Info.plist to Free App Info.plist

Commit:
Remove bogus entitlements

Commit:
Remove Generated.xcconfig

Commit:
Codesign automatic

Commit:
Modify scheme so that paid app works

Commit:
Audit project.pbxproj

Commit:
Remove unused configs

Commit:
share one info.plist

Commit:
pipe flavor for flutter run to work

Commit:
Add devicelab tests

jmagman commented:
I'll review once this is rebased on https://github.com/flutter/flutter/pull/117352, looks like this PR also includes that?

a-wallen commented:
@stuartmorgan it looks like `flutter install` on all desktop devices doesn't do anything at the moment. Is there supposed to be a concrete implementation on `macOS`?

jmagman commented:
> @stuartmorgan it looks like `flutter install` on all desktop devices doesn't do anything at the moment

@jonahwilliams may remember https://github.com/flutter/flutter/pull/26593/files#r249918739

a-wallen commented on dev/devicelab/bin/tasks/flavors_test_macos.dart:
```diff
@@ -0,0 +1,18 @@
+  deviceOperatingSystem = DeviceOperatingSystem.macos;
+  await task(() async {
+    await createFlavorsTest().call();
+    await createIntegrationTestFlavorsTest().call();
```
The file is missing `install` tests, see `dev/devicelab/bin/tasks/flavors_test.dart`.

stuartmorgan commented:
> How about adding for some clarity?

Saying it's "currently unavailable" seems even less clear IMO, because it suggests there's some missing functionality that we plan to implement. AFAIK there's no reason for us to ever implement it (unless we support remote cross-desktop devices at some point).

jonahwilliams commented:
flutter install on mobile deivces copies the APK/IPA to the target device. When developing on the desktop, the host device is the target device, so there is nothing to install. We could make flutter install on desktop simply no-op and return `0`, 🤷 .

If we added support for developing linux/macos/windows with a different target and host machine, we'd almost certainly need to implement a different device type/discovery. In that case `flutter install` wouldn't be a no-op, but since that doesn't exist I wouldn't worry about it.

Commit:
Error if host and target device are same for flutter install desktop

a-wallen commented:
Thanks @jonahwilliams and @stuartmorgan! Please let me know if https://github.com/flutter/flutter/pull/118421/commits/75aaed76816313a32ecc078b38ae74ccda6c48f9 adequately addresses those comments and if there's a better way to check whether the host and target platform are the same.

a-wallen commented:
Looks like the PR that adds a new target in `dev/integration_tests/flavors` is blocked in CI on `Google Testing` without any progress (see b/261725731).

jmagman commented:
> Review on this PR is blocked until #117352, which lands some commits in this PR ([c22f1cc](https://github.com/flutter/flutter/commit/c22f1cc43541c06948441cd4fc51819060c365f1) - [727ee2c](https://github.com/flutter/flutter/commit/727ee2c1135cc59ea2afd36cb659b9695d329b19)) is unblocked on `Google Testing` (see b/261725731).

Merge https://github.com/flutter/flutter/pull/117352 manually, it shouldn't impact g3.

jmagman commented on dev/integration_tests/flavors/macos/Runner.xcodeproj/project.pbxproj:
```diff
@@ -651,6 +651,7 @@
 					"@executable_path/../Frameworks",
 				);
 				PRODUCT_BUNDLE_IDENTIFIER = com.yourcompany.flavors.free;
+				PRODUCT_FLAVOR = free;
```
Why was this needed?  The tool should be parsing the flavor from the configuration, I think?
macOS flavors should work with the same setup as iOS: https://docs.flutter.dev/deployment/flavors#creating-flavors-in-ios

a-wallen replied:
It probably isn't needed, but I followed the way that the `iOS` project was set up.

https://github.com/flutter/flutter/blob/997d43618dc73b2d01a86adaccf9dc9672953c03/dev/integration_tests/flavors/ios/Runner.xcodeproj/project.pbxproj#L474

https://github.com/flutter/flutter/blob/997d43618dc73b2d01a86adaccf9dc9672953c03/dev/integration_tests/flavors/ios/Runner/Info-Free.plist#L24

https://github.com/flutter/flutter/blob/997d43618dc73b2d01a86adaccf9dc9672953c03/dev/integration_tests/flavors/ios/Runner/AppDelegate.m#L16-L19

jmagman replied:
Ahh, okay that's the thing that's being used to test that the flavors are hooked up correctly.

So you do need it:
https://github.com/flutter/flutter/pull/117352/files#diff-497597318ad21439d629a57213312c930c1902042dfb2dc31373f0b42fe4d45bR27
https://github.com/flutter/flutter/pull/117352/files#diff-147b89acf8a22478c244cca300415a21dad69ebfdb0ac61051c12b76b1412100R23-R24

Did that test pass before you added this build setting?

jmagman commented on packages/flutter_tools/lib/src/commands/install.dart:
```diff
@@ -50,6 +51,11 @@ class InstallCommand extends FlutterCommand with DeviceBasedDevelopmentArtifacts
       throwToolExit('No target device found');
     }
+
+    if (const LocalPlatform().operatingSystem.toLowerCase() == device?.name.toLowerCase()) {
```
This logic looks very suspect to detect a desktop device.  Can the device itself say whether it supports installation?

https://github.com/flutter/flutter/blob/997d43618dc73b2d01a86adaccf9dc9672953c03/packages/flutter_tools/lib/src/device.dart#L566

a-wallen replied:
https://github.com/flutter/flutter/pull/118421/commits/d12e9a651a02ad077a1981d44efb13845577a915 should have those revisions.

jmagman commented on packages/flutter_tools/lib/src/macos/application_package.dart:
```diff
@@ -156,23 +156,29 @@ class BuildableMacOSApp extends MacOSApp {
     }
+
+    String bundleDirectory = sentenceCase(buildInfo.mode.name);
+    if(buildInfo.flavor != null) {
+      bundleDirectory += ' ${sentenceCase(buildInfo.flavor!)}';
```
Avoid the bang by promoting a local.
```suggestion
    final String? flavor = buildInfo.flavor;
    if (flavor != null) {
      bundleDirectory += ' ${sentenceCase(flavor)}';
```

Wow I'd never seen this:
https://github.com/flutter/flutter/blob/997d43618dc73b2d01a86adaccf9dc9672953c03/packages/flutter_tools/lib/src/xcode_project.dart#L590-L594
I was going to say we shouldn't be reverse engineering these paths, we should let Xcode tell us to where it's building the app via build settings like `TARGET_BUILD_DIR`.  However the logic that's already here before this PR is doing something similar...

a-wallen committed:
Avoid bang (!) by promoting a local.

Co-authored-by: Jenn Magder <magder@google.com>

Commit:
Add supportsInstall property

Commit:
Override  in test classes

Commit:
Add install test on macOS

jmagman commented on .ci.yaml:
```diff
@@ -2609,6 +2609,20 @@ targets:
       - bin/**
       - .ci.yaml
 
+  - name: Mac flavors_test_macos
```
I tried to run this locally and it got stuck on:
```
[flavors_test_macos] [STDOUT] stdout: 02:22 +0: loading /Users/m/Projects/flutter/dev/integration_tests/flavors/integration_test/integration_test.dart
[flavors_test_macos] [STDOUT] stdout: 02:23 +0: loading /Users/m/Projects/flutter/dev/integration_tests/flavors/integration_test/integration_test.dart
[flavors_test_macos] [STDOUT] stdout: 02:24 +0: loading /Users/m/Projects/flutter/dev/integration_tests/flavors/integration_test/integration_test.dart
[flavors_test_macos] [STDOUT] stdout: 02:25 +0: loading /Users/m/Projects/flutter/dev/integration_tests/flavors/integration_test/integration_test.dart
[flavors_test_macos] [STDOUT] stdout: 02:26 +0: loading /Users/m/Projects/flutter/dev/integration_tests/flavors/integration_test/integration_test.dart
```

Did you see this locally when you ran it?  I'm not sure why it wouldn't be loading in my environment...

a-wallen replied:
Running `dart ./bin/run.dart -t flavors_test_macos` hangs on 

```
[flavors_test_macos] [STDOUT] The Dart VM service is listening on http://127.0.0.1:60212/-JV7h6ggCrk=/
[flavors_test_macos] Connected to VM server.
[flavors_test_macos] [STDOUT] Running task with a timeout of null.
[flavors_test_macos] [STDOUT] 
[flavors_test_macos] [STDOUT] 
[flavors_test_macos] [STDOUT] ══════════════════╡ ••• Checking running Dart processes ••• ╞═══════════════════
[flavors_test_macos] [STDOUT]
```

I don't think that I'm getting to the same point that you are...

jmagman commented on packages/flutter_tools/lib/src/macos/application_package.dart:
```diff
@@ -156,23 +156,30 @@ class BuildableMacOSApp extends MacOSApp {
       return null;
     }
+
+    String bundleDirectory = sentenceCase(buildInfo.mode.name);
+    final String? flavor = buildInfo.flavor;
+    if (flavor != null) {
+      bundleDirectory += ' ${sentenceCase(flavor)}';
+    }
```
Is there a test for this?

a-wallen replied:
There is now https://github.com/flutter/flutter/pull/118421/commits/8b25cb74e0f894ed4299db387cdfae563766436d 👍

Commit:
Refactor application_package and add tests for package directory

jmagman reviewed (approved):
LGTM
